File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/RegisterTimeTriggerInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.model.InstructionBox;

/** Typed builder for {@code RegisterTimeTrigger} instructions. */
public final class RegisterTimeTriggerInstruction implements InstructionTemplate {

public static final String ACTION = "RegisterTimeTrigger";
  static final String REPEATS_INDEFINITE = "indefinite";

  private final String triggerId;
  private final String authority;
  private final long startMs;
  private final Long periodMs;
  private final Integer repeats;
  private final List<InstructionBox> instructions;
  private final Map<String, String> metadata;
  private final Map<String, String> arguments;

  private RegisterTimeTriggerInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private RegisterTimeTriggerInstruction(
      final Builder builder, final Map<String, String> argumentOrder) {
    this.triggerId = builder.triggerId;
    this.authority = builder.authority;
    this.startMs = builder.startMs;
    this.periodMs = builder.periodMs;
    this.repeats = builder.repeats;
    this.instructions = Collections.unmodifiableList(new ArrayList<>(builder.instructions));
    this.metadata =
        Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(builder.metadata)));
    this.arguments =
        Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(argumentOrder)));
  }

  public String triggerId() {
    return triggerId;
  }

  public String authority() {
    return authority;
  }

  public long startMs() {
    return startMs;
  }

  public Long periodMs() {
    return periodMs;
  }

  public Integer repeats() {
    return repeats;
  }

  public List<InstructionBox> instructions() {
    return instructions;
  }

  public Map<String, String> metadata() {
    return metadata;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.REGISTER;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static RegisterTimeTriggerInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder =
        builder()
            .setTriggerId(require(arguments, "trigger"))
            .setAuthority(require(arguments, "authority"))
            .setStartMs(parsePositiveLong(require(arguments, "start_ms"), "start_ms"))
            .setInstructions(TriggerInstructionUtils.parseInstructions(arguments))
            .setMetadata(TriggerInstructionUtils.extractMetadata(arguments));
    if (arguments.containsKey("period_ms")) {
      builder.setPeriodMs(parsePositiveLong(arguments.get("period_ms"), "period_ms"));
    }
    final Integer repeats = TriggerInstructionUtils.parseRepeats(arguments.get("repeats"));
    if (repeats != null) {
      builder.setRepeats(repeats);
    }
    return new RegisterTimeTriggerInstruction(builder, new LinkedHashMap<>(arguments));
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  private static long parsePositiveLong(final String value, final String field) {
    try {
      final long parsed = Long.parseUnsignedLong(value);
      if (parsed <= 0) {
        throw new IllegalArgumentException(field + " must be greater than zero");
      }
      return parsed;
    } catch (final NumberFormatException ex) {
      throw new IllegalArgumentException(field + " must be an unsigned integer", ex);
    }
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof RegisterTimeTriggerInstruction other)) {
      return false;
    }
    return startMs == other.startMs
        && Objects.equals(triggerId, other.triggerId)
        && Objects.equals(authority, other.authority)
        && Objects.equals(periodMs, other.periodMs)
        && Objects.equals(repeats, other.repeats)
        && instructions.equals(other.instructions)
        && metadata.equals(other.metadata);
  }

  @Override
  public int hashCode() {
    return Objects.hash(triggerId, authority, startMs, periodMs, repeats, instructions, metadata);
  }

  public static final class Builder {
    private String triggerId;
    private String authority;
    private Long startMs;
    private Long periodMs;
    private Integer repeats;
    private final List<InstructionBox> instructions = new ArrayList<>();
    private final Map<String, String> metadata = new LinkedHashMap<>();

    private Builder() {}

    public Builder setTriggerId(final String triggerId) {
      if (triggerId == null || triggerId.isBlank()) {
        throw new IllegalArgumentException("triggerId must not be blank");
      }
      this.triggerId = triggerId;
      return this;
    }

    public Builder setAuthority(final String authority) {
      this.authority =
          org.hyperledger.iroha.android.address.AccountIdLiteral.requireCanonicalI105Address(
              authority, "authority");
      return this;
    }

    public Builder setStartMs(final long startMs) {
      if (startMs <= 0) {
        throw new IllegalArgumentException("startMs must be greater than zero");
      }
      this.startMs = startMs;
      return this;
    }

    public Builder setPeriodMs(final Long periodMs) {
      if (periodMs == null) {
        this.periodMs = null;
        return this;
      }
      if (periodMs <= 0) {
        throw new IllegalArgumentException("periodMs must be greater than zero when provided");
      }
      this.periodMs = periodMs;
      return this;
    }

    public Builder setRepeats(final Integer repeats) {
      if (repeats == null) {
        this.repeats = null;
        return this;
      }
      if (repeats <= 0) {
        throw new IllegalArgumentException("repeats must be greater than zero when provided");
      }
      this.repeats = repeats;
      return this;
    }

    public Builder addInstruction(final InstructionBox instruction) {
      instructions.add(Objects.requireNonNull(instruction, "instruction"));
      return this;
    }

    public Builder setInstructions(final List<InstructionBox> newInstructions) {
      instructions.clear();
      if (newInstructions != null) {
        newInstructions.forEach(this::addInstruction);
      }
      return this;
    }

    public Builder putMetadata(final String key, final String value) {
      metadata.put(Objects.requireNonNull(key, "metadata key"), Objects.requireNonNull(value, "metadata value"));
      return this;
    }

    public Builder setMetadata(final Map<String, String> entries) {
      metadata.clear();
      if (entries != null) {
        entries.forEach(this::putMetadata);
      }
      return this;
    }

    public RegisterTimeTriggerInstruction build() {
      if (triggerId == null) {
        throw new IllegalStateException("triggerId must be set");
      }
      if (authority == null) {
        throw new IllegalStateException("authority must be set");
      }
      if (startMs == null) {
        throw new IllegalStateException("startMs must be set");
      }
      if (instructions.isEmpty()) {
        throw new IllegalStateException("at least one instruction must be provided");
      }
      return new RegisterTimeTriggerInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put("trigger", triggerId);
      args.put("authority", authority);
      args.put("start_ms", Long.toUnsignedString(startMs));
      if (periodMs != null) {
        args.put("period_ms", Long.toUnsignedString(periodMs));
      }
      args.put(
          "repeats",
          repeats == null ? REPEATS_INDEFINITE : Integer.toUnsignedString(repeats));
      TriggerInstructionUtils.appendInstructions(instructions, args);
      TriggerInstructionUtils.appendMetadata(metadata, args);
      return args;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/RegisterTimeTriggerInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

import org.hyperledger.iroha.sdk.core.model.InstructionBox

/** Typed builder for `RegisterTimeTrigger` instructions. */
class RegisterTimeTriggerInstruction private constructor(
    @JvmField val triggerId: String,
    @JvmField val authority: String,
    @JvmField val startMs: Long,
    @JvmField val periodMs: Long?,
    @JvmField val repeats: Int?,
    @JvmField val instructions: List<InstructionBox>,
    @JvmField val metadata: Map<String, String>,
    override val arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind get() = InstructionKind.REGISTER

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RegisterTimeTriggerInstruction) return false
        return startMs == other.startMs
            && triggerId == other.triggerId
            && authority == other.authority
            && periodMs == other.periodMs
            && repeats == other.repeats
            && instructions == other.instructions
            && metadata == other.metadata
    }

    override fun hashCode(): Int {
        var result = triggerId.hashCode()
        result = 31 * result + authority.hashCode()
        result = 31 * result + startMs.hashCode()
        result = 31 * result + (periodMs?.hashCode() ?: 0)
        result = 31 * result + (repeats?.hashCode() ?: 0)
        result = 31 * result + instructions.hashCode()
        result = 31 * result + metadata.hashCode()
        return result
    }

    class Builder internal constructor() {
        private var triggerId: String? = null
        private var authority: String? = null
        private var startMs: Long? = null
        private var periodMs: Long? = null
        private var repeats: Int? = null
        private val instructions: MutableList<InstructionBox> = mutableListOf()
        private val metadata: MutableMap<String, String> = linkedMapOf()

        fun setTriggerId(triggerId: String) = apply {
            require(triggerId.isNotBlank()) { "triggerId must not be blank" }
            this.triggerId = triggerId
        }

        fun setAuthority(authority: String) = apply {
            require(authority.isNotBlank()) { "authority must not be blank" }
            this.authority = authority
        }

        fun setStartMs(startMs: Long) = apply {
            require(startMs > 0) { "startMs must be greater than zero" }
            this.startMs = startMs
        }

        fun setPeriodMs(periodMs: Long?) = apply {
            if (periodMs != null) {
                require(periodMs > 0) { "periodMs must be greater than zero when provided" }
            }
            this.periodMs = periodMs
        }

        fun setRepeats(repeats: Int?) = apply {
            if (repeats != null) {
                require(repeats > 0) { "repeats must be greater than zero when provided" }
            }
            this.repeats = repeats
        }

        fun addInstruction(instruction: InstructionBox) = apply {
            instructions.add(requireNotNull(instruction) { "instruction" })
        }

        fun setInstructions(newInstructions: List<InstructionBox>?) = apply {
            instructions.clear()
            newInstructions?.forEach { addInstruction(it) }
        }

        fun putMetadata(key: String, value: String) = apply {
            metadata[requireNotNull(key) { "metadata key" }] = requireNotNull(value) { "metadata value" }
        }

        fun setMetadata(entries: Map<String, String>?) = apply {
            metadata.clear()
            entries?.forEach { (k, v) -> putMetadata(k, v) }
        }

        fun build(): RegisterTimeTriggerInstruction {
            val tid = checkNotNull(triggerId) { "triggerId must be set" }
            val auth = checkNotNull(authority) { "authority must be set" }
            val sm = checkNotNull(startMs) { "startMs must be set" }
            check(instructions.isNotEmpty()) { "at least one instruction must be provided" }
            val args = buildCanonicalArguments(tid, auth, sm, periodMs, repeats, instructions, metadata)
            return RegisterTimeTriggerInstruction(
                tid, auth, sm, periodMs, repeats, instructions.toList(), metadata.toMap(), args,
            )
        }

        private fun buildCanonicalArguments(
            triggerId: String,
            authority: String,
            startMs: Long,
            periodMs: Long?,
            repeats: Int?,
            instructions: List<InstructionBox>,
            metadata: Map<String, String>,
        ): Map<String, String> = buildMap {
            put("action", ACTION)
            put("trigger", triggerId)
            put("authority", authority)
            put("start_ms", java.lang.Long.toUnsignedString(startMs))
            if (periodMs != null) {
                put("period_ms", java.lang.Long.toUnsignedString(periodMs))
            }
            put(
                "repeats",
                if (repeats == null) REPEATS_INDEFINITE
                else Integer.toUnsignedString(repeats),
            )
            TriggerInstructionUtils.appendInstructions(instructions, this)
            TriggerInstructionUtils.appendMetadata(metadata, this)
        }
    }

    companion object {
        const val ACTION: String = "RegisterTimeTrigger"
        @JvmField val REPEATS_INDEFINITE: String = "indefinite"

        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): RegisterTimeTriggerInstruction {
            val triggerId = requireArg(arguments, "trigger")
            val authority = requireArg(arguments, "authority")
            val startMs = parsePositiveLong(requireArg(arguments, "start_ms"), "start_ms")
            val instructions = TriggerInstructionUtils.parseInstructions(arguments)
            val metadata = TriggerInstructionUtils.extractMetadata(arguments)
            val periodMs = arguments["period_ms"]?.let { parsePositiveLong(it, "period_ms") }
            val repeats = TriggerInstructionUtils.parseRepeats(arguments["repeats"])
            return RegisterTimeTriggerInstruction(
                triggerId = triggerId,
                authority = authority,
                startMs = startMs,
                periodMs = periodMs,
                repeats = repeats,
                instructions = instructions,
                metadata = metadata,
                arguments = LinkedHashMap(arguments),
            )
        }

        private fun requireArg(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrBlank()) { "Instruction argument '$key' is required" }
            return value
        }

        private fun parsePositiveLong(value: String, field: String): Long = try {
            val parsed = java.lang.Long.parseUnsignedLong(value)
            if (parsed <= 0) throw IllegalArgumentException("$field must be greater than zero")
            parsed
        } catch (ex: NumberFormatException) {
            throw IllegalArgumentException("$field must be an unsigned integer", ex)
        }
    }
}
